clients/drcachesim/drpt2trace/pt2ir.cpp

```diff
@@ -270,150 +414,188 @@ pt2ir_t::convert(IN const uint8_t *pt_data, IN size_t pt_data_size, INOUT drir_t
         return PT2IR_CONV_ERROR_RAW_TRACE_TOO_LARGE;
     }
 
-    /* The Libipt decoder requires a fixed-size data block to be set for decoding before
-     * creation and does not offer an API to extend or modify the data being decoded. To
-     * decode multiple data chunks using a shared decoder, a workaround is needed. First,
-     * initialize an empty data block as a buffer. Next, each time we want to decode a
-     * chunk, reset the buffer and copy the data into it, and then set the decoder's
-     * decode position to the chunk's initial position.
-     */
+    /* Append the new data to the buffer. */
+    memcpy(pt_raw_buffer_.get() + pt_raw_buffer_data_size_, pt_data, pt_data_size);
+    pt_raw_buffer_data_size_ += pt_data_size;
+
+    if (!pt_decoder_has_sync_) {
+        /* Sync decoder to next Packet Stream Boundary (PSB) packet, and then
+         * decode instructions. If there is no PSB packet, the decoder will be synced
+         * to the end of the trace. If an error occurs, we will call dx_decoding_error
+         * to print the error information.
+         * What are PSB packets? Below answer is quoted from Intel 64 and IA-32
+         * Architectures Software Developer’s Manual 32.1.1.1 Packet Summary:
+         * “Packet Stream Boundary (PSB) packets: PSB packets act as ‘heartbeats’ that are
+         * generated at regular intervals (e.g., every 4K trace packet bytes). These
+         * packets allow the packet decoder to find the packet boundaries within the
+         * output data stream; a PSB packet should be the first packet that a decoder
+         * looks for when beginning to decode a trace.”
+         */
+        pt_instr_status_ = pt_insn_sync_forward(pt_instr_decoder_);
+        if (pt_instr_status_ < 0) {
+            dx_decoding_error(pt_instr_status_, "sync instruction decoder error", 0);
+            return PT2IR_CONV_ERROR_SYNC_PACKET;
+        }
+        pt_decoder_has_sync_ = true;
+    }
 
-    /* Reset the raw buffer and copy the new data to the buffer. */
-    memset(pt_raw_buffer_.get(), 0, pt_raw_buffer_size_);
-    memcpy(pt_raw_buffer_.get(), pt_data, pt_data_size);
-    pt_raw_buffer_data_size_ = pt_data_size;
+    /* TODO i#5505: We only decode a portion of the final packet of the current chunk,
+     * with the remaining part decoded when the next chunk arrives. This results in each
+     * initial chunk missing some instructions at the end, the final chunk having some
+     * extraneous instructions, and the middle chunks' decoding outcomes containing
+     * instructions from the previous chunk at the beginning and missing some at the end.
+     * This approach is acceptable for decoding every syscall's trace since they are
+     * surrounded by noise instructions and the missing instructions only impact these
+     * noise instructions. However, it's important not to employ pt2ir to decode other
+     * discontinuous chunks that demand a precision instruction list.
```

Comment:
So if we want precise decoding, we cannot use the streaming decoder and have the turn off the optimization?

Reply:
Yes.